src/app/navigation/types.ts

```diff
@@ -6,32 +6,44 @@ import { AnyAction } from '../types';
 export type State = Location;
 
 type RouteParams<R> = R extends Route<infer P> ? P : never;
+type RouteState<R> = R extends Route<any, infer S> ? S : never;
+
 type UnionRouteMatches<R> = R extends AnyRoute ? Match<R> : never;
+type UnionHistoryActions<R> = R extends AnyRoute ? HistoryAction<R> : never;
 
-interface MatchWithParams<R extends AnyRoute> {
+interface MatchWithoutParams<R extends AnyRoute> {
   route: R;
+}
+export interface MatchWithParams<R extends AnyRoute> extends MatchWithoutParams<R> {
   params: RouteParams<R>;
 }
-interface MatchWithoutParams<R extends AnyRoute> {
-  route: R;
+export interface MatchWithState<R extends AnyRoute> extends MatchWithoutParams<R> {
+  state?: RouteState<R>;
 }
 
 export type GenericMatch = MatchWithParams<AnyRoute> | MatchWithoutParams<AnyRoute>;
 
-export type Match<R extends AnyRoute> = RouteParams<R> extends undefined
-  ? MatchWithoutParams<R> | MatchWithParams<R>
-  : MatchWithParams<R>;
+export type Match<R extends AnyRoute> =
+  (RouteParams<R> extends undefined
+    ? MatchWithoutParams<R> | MatchWithParams<R>
+    : MatchWithParams<R>)
+  & (RouteState<R> extends undefined
+    ? {}
+    : MatchWithState<R>);
+
+export type HistoryAction<R extends AnyRoute> = Match<R> & {
+  method: 'push' | 'replace';
+};
 
-export type historyActions =
-  {method: 'push', url: string} |
-  {method: 'replace', url: string};
+export type AnyHistoryAction = UnionHistoryActions<AnyRoute>;
 
 export type reducer = (state: State, action: AnyAction) => State;
 
-export interface Route<Params> {
+// @ts-ignore: 'S' is declared but its value is never read.
```

Comment:
Why is `S` never used?

Reply:
the shape of `state` isn't used in defining the members of the route object, but it is a defining characteristic of a route, its used in other types to make sure the state you pass with the route matches the state the route was defined with

Reply:
eg: https://github.com/openstax/rex-web/pull/32/files#diff-558360260e96a66c6b633c30d170dd6bR21